Multigrid evolve test fails to compile

Seems it missed the fix for per-level input/output files in some places. Putting it here so we don't forget to fix it for 1.0.

```
/home/mathiasw/quda/release/1.0.x/tests/multigrid_evolve_test.cpp: In function ‘int main(int, char**)’:
/home/mathiasw/quda/release/1.0.x/tests/multigrid_evolve_test.cpp:829:59: error: cannot convert ‘char (*)[256]’ to ‘char*’ for argument ‘1’ to ‘int sprintf(char*, const char*, ...)’
         sprintf(mg_param.vec_outfile, "dump_step_%d", step);
                                                           ^
In file included from /home/mathiasw/quda/release/1.0.x/tests/multigrid_evolve_test.cpp:7:0:
/home/mathiasw/quda/release/1.0.x/include/util_quda.h:103:42: warning: format ‘%s’ expects argument of type ‘char*’, but argument 3 has type ‘char (*)[256]’ [-Wformat=]
     sprintf(getPrintBuffer(), __VA_ARGS__);   \
                                          ^
/home/mathiasw/quda/release/1.0.x/tests/multigrid_evolve_test.cpp:830:9: note: in expansion of macro ‘warningQuda’
         warningQuda("Solver failed to converge within max iteration count - dumping null vectors to %s",
         ^~~~~~~~~~~
/home/mathiasw/quda/release/1.0.x/tests/multigrid_evolve_test.cpp:834:38: error: ‘vec_outfile’ was not declared in this scope
         strcpy(mg_param.vec_outfile, vec_outfile); // restore output file name
                                      ^~~~~~~~~~~
/home/mathiasw/quda/release/1.0.x/tests/multigrid_evolve_test.cpp:834:38: note: suggested alternative: ‘mg_vec_outfile’
         strcpy(mg_param.vec_outfile, vec_outfile); // restore output file name
                                      ^~~~~~~~~~~
                                      mg_vec_outfile```

--- a/lib/blas_quda.cu
+++ b/lib/blas_quda.cu
@@ -403,7 +403,7 @@ namespace quda {
 
 #if QUDA_PRECISION & 8
             if (x.Nspin() == 4) {
-              const int M = 6;
+              const int M = 12;
               nativeBlas<double2, char4, double2, M, Functor, writeX, writeY, writeZ, writeW, writeV>(
                   a, b, c, x, y, z, w, v, x.Volume());
             } else if (x.Nspin() == 1) {

--- a/lib/contract.cu
+++ b/lib/contract.cu
@@ -4,6 +4,7 @@
 #include <blas_quda.h>
 
 #include <contract_quda.h>
+#include <jitify_helper.cuh>
 #include <kernels/contraction.cuh>
 
 namespace quda {

--- a/lib/multi_blas_quda.cu
+++ b/lib/multi_blas_quda.cu
@@ -457,6 +457,30 @@ namespace quda {
             errorQuda("QUDA_PRECISION=%d does not enable precision %d", QUDA_PRECISION, x[0]->Precision());
 #endif
 
+          } else if (x[0]->Precision() == QUDA_QUARTER_PRECISION) {
+
+#if QUDA_PRECISION & 1
+            if (x[0]->Nspin() == 4) {
+#if defined(GPU_WILSON_DIRAC) || defined(GPU_DOMAIN_WALL_DIRAC)
+              const int M = 12;
+              multiBlas<NXZ, double2, char4, double2, M, Functor, write>(a, b, c, x, y, z, w, x[0]->Volume());
+#else
+              errorQuda("blas has not been built for Nspin=%d fields", x[0]->Nspin());
+#endif
+
+            } else if (x[0]->Nspin() == 1) {
+
+#if defined(GPU_STAGGERED_DIRAC)
+              const int M = 3;
+              multiBlas<NXZ, double2, char2, double2, M, Functor, write>(a, b, c, x, y, z, w, x[0]->Volume());
+#else
+              errorQuda("blas has not been built for Nspin=%d fields", x[0]->Nspin());
+#endif
+            }
+#else
+            errorQuda("QUDA_PRECISION=%d does not enable precision %d", QUDA_PRECISION, x[0]->Precision());
+#endif
+
           } else {
             errorQuda("Not implemented for this precision combination %d %d", x[0]->Precision(), y[0]->Precision());
           }

--- a/lib/reduce_quda.cu
+++ b/lib/reduce_quda.cu
@@ -420,7 +420,7 @@ namespace quda {
 #if defined(GPU_MULTIGRID)
             const int M = 12; // determines how much work per thread to do
             value
-                = nativeReduce<doubleN, ReduceType, float2, char2, char2, M, Reducer, writeX, writeY, writeZ, writeW, writeV>(
+                = nativeReduce<doubleN, ReduceType, float2, short2, short2, M, Reducer, writeX, writeY, writeZ, writeW, writeV>(
                     a, b, x, y, z, w, v, y.Volume());
 #else
             errorQuda("blas has not been built for Nspin=%d fields", x.Nspin());
@@ -575,6 +575,32 @@ namespace quda {
             errorQuda("QUDA_PRECISION=%d does not enable precision %d", QUDA_PRECISION, x.Precision());
 #endif
 
+          } else if (x.Precision() == QUDA_QUARTER_PRECISION) {
+
+#if QUDA_PRECISION & 1
+            if (x.Nspin() == 4) { // wilson
+#if defined(GPU_WILSON_DIRAC) || defined(GPU_DOMAIN_WALL_DIRAC)
+              const int M = 12; // determines how much work per thread to do
+              value = nativeReduce<doubleN, ReduceType, double2, char4, double2, M, Reducer, writeX, writeY, writeZ,
+                  writeW, writeV>(a, b, x, y, z, w, v, x.Volume());
+#else
+              errorQuda("blas has not been built for Nspin=%d fields", x.Nspin());
+#endif
+            } else if (x.Nspin() == 1) { // staggered
+#ifdef GPU_STAGGERED_DIRAC
+              const int M = 3; // determines how much work per thread to do
+              value = nativeReduce<doubleN, ReduceType, double2, char2, double2, M, Reducer, writeX, writeY, writeZ,
+                  writeW, writeV>(a, b, x, y, z, w, v, x.Volume());
+#else
+              errorQuda("blas has not been built for Nspin=%d fields", x.Nspin());
+#endif
+            } else {
+              errorQuda("ERROR: nSpin=%d is not supported\n", x.Nspin());
+            }
+#else
+            errorQuda("QUDA_PRECISION=%d does not enable precision %d", QUDA_PRECISION, x.Precision());
+#endif
+
           } else {
             errorQuda("Not implemented for this precision combination %d %d", x.Precision(), z.Precision());
           }
@@ -600,6 +626,33 @@ namespace quda {
 #ifdef GPU_STAGGERED_DIRAC
               const int M = 3;
               value = nativeReduce<doubleN, ReduceType, float2, short2, float2, M, Reducer, writeX, writeY, writeZ,
+                  writeW, writeV>(a, b, x, y, z, w, v, x.Volume());
+#else
+              errorQuda("blas has not been built for Nspin=%d fields", x.Nspin());
+#endif
+            } else {
+              errorQuda("ERROR: nSpin=%d is not supported\n", x.Nspin());
+            }
+            blas::bytes
+                += Reducer<ReduceType, double2, double2>::streams() * (unsigned long long)x.Volume() * sizeof(float);
+#else
+            errorQuda("QUDA_PRECISION=%d does not enable precision %d", QUDA_PRECISION, x.Precision());
+#endif
+
+          } else if (x.Precision() == QUDA_QUARTER_PRECISION) {
+#if QUDA_PRECISION & 1
+            if (x.Nspin() == 4) { // wilson
+#if defined(GPU_WILSON_DIRAC) || defined(GPU_DOMAIN_WALL_DIRAC)
+              const int M = 6;
+              value = nativeReduce<doubleN, ReduceType, float4, char4, float4, M, Reducer, writeX, writeY, writeZ,
+                  writeW, writeV>(a, b, x, y, z, w, v, x.Volume());
+#else
+              errorQuda("blas has not been built for Nspin=%d fields", x.Nspin());
+#endif
+            } else if (x.Nspin() == 1) { // staggered
+#ifdef GPU_STAGGERED_DIRAC
+              const int M = 3;
+              value = nativeReduce<doubleN, ReduceType, float2, char2, float2, M, Reducer, writeX, writeY, writeZ,
                   writeW, writeV>(a, b, x, y, z, w, v, x.Volume());
 #else
               errorQuda("blas has not been built for Nspin=%d fields", x.Nspin());

--- a/tests/multigrid_evolve_test.cpp
+++ b/tests/multigrid_evolve_test.cpp
@@ -826,12 +826,18 @@ int main(int argc, char **argv)
       invertQuda(spinorOut, spinorIn, &inv_param);
 
       if (inv_param.iter == inv_param.maxiter) {
-        sprintf(mg_param.vec_outfile, "dump_step_%d", step);
+        char vec_outfile[QUDA_MAX_MG_LEVEL][256];
+        for (int i=0; i<mg_param.n_level; i++) {
+          strcpy(vec_outfile[i], mg_param.vec_outfile[i]);
+          sprintf(mg_param.vec_outfile[i], "dump_step_%d", step);
+        }
         warningQuda("Solver failed to converge within max iteration count - dumping null vectors to %s",
-                    mg_param.vec_outfile);
+                    mg_param.vec_outfile[0]);
 
         dumpMultigridQuda(mg_preconditioner, &mg_param);
-        strcpy(mg_param.vec_outfile, vec_outfile); // restore output file name
+        for (int i=0; i<mg_param.n_level; i++) {
+          strcpy(mg_param.vec_outfile[i], vec_outfile[i]); // restore output file name
+        }
       }
 
       freeGaugeQuda();

